### PR TITLE
Wrangler performance optimizations

### PR DESCRIPTION
**Pull request by tivv**
* Cache of lowercase column names in Row for lookup. During Row->Structured records currently it's n^2 of toLowerCase, this should make it n for toLowerCase
 * javax.script.SimpleScriptContext constructor is pretty expensive as it unconditionally creates a number of Buffered readers/writer. This change caches it as ThreadLocal

### CODE REVIEW

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java [hunk 1]
```diff
@@ -141,6 +156,9 @@ public Row setValue(int idx, Object value) {
   public Row add(String name, Object value) {
     columns.add(name);
     values.add(value);
+    if (nameIndex != null && name != null) {
+      nameIndex.put(name.toLowerCase(Locale.ENGLISH), columns.size());
```
Comment (seanzhougoogle) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java]: `columns.size()-1` ?

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java [hunk 2]
```diff
@@ -150,8 +168,11 @@ public Row add(String name, Object value) {
    * @param idx for which the value and column are removed.
    */
   public Row remove(int idx) {
-    columns.remove(idx);
+    String name = columns.remove(idx);
     values.remove(idx);
+    if (name != null && nameIndex != null) {
+      nameIndex.remove(name.toLowerCase(Locale.ENGLISH));
```
Comment (seanzhougoogle) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java]: remove from the middle will also change the index after it , right ?

File: wrangler-transform/src/main/java/io/cdap/wrangler/Precondition.java [hunk 3]
```diff
@@ -45,14 +49,28 @@ public Precondition(String condition) throws PreconditionException {
     }
   }
 
+  public ScriptContext createContext() {
```
Comment (seanzhougoogle) [inline wrangler-transform/src/main/java/io/cdap/wrangler/Precondition.java:52]: Not familiar with ScriptConext, I saw this part is copied from `CompiledScript.eval`, is it implementation specific ? will it work with later java versions ?

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java (hunk 1 above)
Comment (fernst) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java]: good catch, is this a 0-based or 1-based index?

File: wrangler-transform/src/main/java/io/cdap/wrangler/Precondition.java (hunk 3 above)
Comment (tivv) [reply wrangler-transform/src/main/java/io/cdap/wrangler/Precondition.java:52]: It does not use any private API, so I don't expect problems here

Comment (tivv) [reply wrangler-transform/src/main/java/io/cdap/wrangler/Precondition.java:52]: Also it's not from specific implementation, but from CompiledScript itself

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java (hunk 1 above)
Comment (tivv) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java]: Done

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java (hunk 2 above)
Comment (tivv) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java]: Yes, good catch, fixed.

**Review by seanzhougoogle — APPROVED** [review]

**Pull request merged**